### PR TITLE
ephemeral values updates

### PR DESCRIPTION
**Pull request by robin-norwood**

### CODE REVIEW

File: website/docs/language/values/locals.mdx [hunk 1]
```diff
@@ -54,6 +54,37 @@ locals {
 }
 ```
 
+## Ephemeral values
+
+Local values will implicitly become ephemeral if an ephemeral value is
```
Comment (rkoron007) [inline website/docs/language/values/locals.mdx:59]:
```suggestion
Local values implicitly become ephemeral if you reference an ephemeral value when
```

File: website/docs/language/resources/ephemeral.mdx [hunk 2]
```diff
@@ -0,0 +1,137 @@
+---
+page_title: ephemeral resources configuration reference
+description: Learn about ephemeral resource blocks that you can specify in Terraform configurations. Ephemeral resource blocks represent temporary resources that Terraform does not store in state.
+---
+
+# Ephemeral resource configuration reference
+
+This topic provides reference information for the `moved` block. 
+
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:9]: Can you add a beta warning to the top of this page? Something like:
```suggestion

-> Ephemeral resources are in public beta and available in Terraform v1.1.0 and later. The workflows of ephemeral resources are subject to change.

```

File: website/docs/language/resources/ephemeral.mdx [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+ephemeral resources, such as a temporary password or connection to another
+system.
+
+<!-- After page exists, add: For information about how Terraform manages ephemeral resources, refer to
+[Epehemeral Values](/terraform/language/epehemeral-values). -->
+
+## Ephemeral Resource Syntax
+
+An `ephemeral` block declares an ephemeral resource of a specific type with a
+specific local name, much like a `resource` block. Terraform uses the name when
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:23]:
```suggestion
specific local name, much like a `resource` block. Terraform uses an ephemeral resource's name to
```

File: website/docs/language/resources/ephemeral.mdx [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+referring to the ephemeral resource in the same module, but it has no meaning
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:24]:
```suggestion
refer to that resource in the same module, but an ephemeral resource's name has no meaning
```

File: website/docs/language/resources/ephemeral.mdx [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+outside that module's scope.
+
+### Ephemeral Resource Arguments
+
+Most of the arguments within the body of an `ephemeral` block are specific to
+the selected ephemeralresource type. The ephemeral resource type's documentation
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:30]:
```suggestion
the ephemeral resource you are defining. The ephemeral resource type's documentation in the [Registry](https://registry.terraform.io/) 
```

File: website/docs/language/resources/ephemeral.mdx [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+lists which arguments are available and how their values should be formatted.
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:31]:
```suggestion
lists which arguments are available and how you should format your resource's values.
```

File: website/docs/language/resources/ephemeral.mdx [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+[Meta-arguments](#meta-arguments) are defined by Terraform
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:33]: I actually think we can remove this, you cover it soon enough below
```suggestion
```

File: website/docs/language/resources/ephemeral.mdx [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+and apply across all ephemeral resource types.
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:34]: To go with above suggestion:
```suggestion
```

File: website/docs/language/resources/ephemeral.mdx [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+lists which arguments are available and how their values should be formatted.
+
+[Meta-arguments](#meta-arguments) are defined by Terraform
+and apply across all ephemeral resource types. 
+
+### Documentation for Resource Types
+
+As with resources and data sources, each provider in the [Terraform
+Registry](https://registry.terraform.io/browse/providers) includes documentation
+for ephemeral resources it supports, if any.
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:40]:
```suggestion
for the ephemeral resources it supports, if any.
```

File: website/docs/language/resources/ephemeral.mdx [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+## Meta-Arguments
+
+The Terraform language defines the following meta-arguments, which can be used
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:44]:
```suggestion
The Terraform language defines the following meta-arguments, which you can use
```

File: website/docs/language/resources/ephemeral.mdx [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+with any ephemeral resource type to change the behavior of resources. These
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:45]:
```suggestion
with any ephemeral resource type to change the behavior of those resources. These
```

File: website/docs/language/resources/ephemeral.mdx [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+meta-arguments work the same way for resources, data sources, and ephemeral
+resources:
+
+- [`depends_on`, for specifying hidden dependencies](/terraform/language/meta-arguments/depends_on)
+- [`count`, for creating multiple resource instances according to a count](/terraform/language/meta-arguments/count)
+- [`for_each`, to create multiple instances according to a map, or set of strings](/terraform/language/meta-arguments/for_each)
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:51]:
```suggestion
- [`for_each`, to create multiple instances according to a map or set of strings](/terraform/language/meta-arguments/for_each)
```

File: website/docs/language/resources/ephemeral.mdx [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+- [`provider`, for selecting a non-default provider configuration](/terraform/language/meta-arguments/resource-provider)
+- [`lifecycle`, for lifecycle customizations](/terraform/language/meta-arguments/lifecycle)
+
+Ephemeral resources to not support the `provisioner` meta-argument.
+
+## Ephemeral resource lifecycle
+
+The lifecycle of ephemeral resources is different resources and data sources.
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:59]:
```suggestion
The lifecycle of ephemeral resources differs from other resources and data sources.
```

File: website/docs/language/resources/ephemeral.mdx [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+When provisioning ephemeral resources, Terraform will open the resource,
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:60]:
```suggestion
When provisioning ephemeral resources, Terraform opens the resource and optionally renews it one or more times before closing it.
```

File: website/docs/language/resources/ephemeral.mdx [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+optionally renew it one or more times, and then close it.
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:61]:
```suggestion
```

File: website/docs/language/resources/ephemeral.mdx [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:62]: Can you introduce this list with an opening sentence? Like is this the traditional lifecycle or is it branching?

File: website/docs/language/resources/ephemeral.mdx [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+1. When Terraform needs to access the ephemeral resource result, it will open
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:63]:
```suggestion
1. If Terraform needs to access an ephemeral resource's result, it opens
```

File: website/docs/language/resources/ephemeral.mdx [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+the ephemeral resource. For example, opening a vault secret ephemeral resource
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:64]:
```suggestion
that ephemeral resource. For example, if Terraform opens an ephemeral resource for a Vault secret, the Vault provider obtains a lease and returns a secret. 
```

File: website/docs/language/resources/ephemeral.mdx [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+means the Vault provider will obtain the lease and return the secret.
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:65]:
```suggestion
```

File: website/docs/language/resources/ephemeral.mdx [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+
+1. If Terraform needs access to the ephemeral resource for longer than the
+expiration time enforced by the remote system, Terraform will ask the provider
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:68]: What is the "expiration time enforced by the remote system"? Is this a standard time? Or something dictated by the provider?
```suggestion
remote system's enforced expiration time, Terraform asks the provider
```

File: website/docs/language/resources/ephemeral.mdx [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+to renew it periodically. For example, renewing a Vault secret ephemeral
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:69]:
```suggestion
to renew it periodically. For example, if Terraform renews a Vault secret ephemeral
```

File: website/docs/language/resources/ephemeral.mdx [hunk 22]
```diff
@@ -0,0 +1,137 @@
...
+resource means the Vault provider will call Vault's lease renewal API endpoint
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:70]:
```suggestion
resource, the Vault provider then calls Vault's lease renewal API endpoint
```

File: website/docs/language/resources/ephemeral.mdx [hunk 23]
```diff
@@ -0,0 +1,137 @@
...
+to extend the expiration time.
+
+1. Once Terraform no longer needs the ephemeral resource, Terraform will close
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:73]:
```suggestion
1. Once Terraform no longer needs an ephemeral resource, Terraform closes
```

File: website/docs/language/resources/ephemeral.mdx [hunk 24]
```diff
@@ -0,0 +1,137 @@
...
+it. This will happen once any providers that depend on the ephemeral resource
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:74]:
```suggestion
it. This happens once the providers that depend on a certain ephemeral resource
```

File: website/docs/language/resources/ephemeral.mdx [hunk 25]
```diff
@@ -0,0 +1,137 @@
...
+have completed all of their work for the current phase. For example, closing a
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:75]:
```suggestion
complete all of their work for the current Terraform run phase. For example, closing a
```

File: website/docs/language/resources/ephemeral.mdx [hunk 26]
```diff
@@ -0,0 +1,137 @@
...
+Vault secret ephemeral resource means the Vault provider will explicitly end the
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:76]:
```suggestion
Vault secret ephemeral resource means the Vault provider explicitly ends the
```

File: website/docs/language/resources/ephemeral.mdx [hunk 27]
```diff
@@ -0,0 +1,137 @@
...
+lease, allowing Vault to immediately revoke the associated credentials.
+
+Terraform will follow these lifecycle steps for each instance of an ephemeral
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:79]: Oh, this sentence is great! Can you move this or a version of it up to the beginning of this list?
```suggestion
Terraform follows these lifecycle steps for each instance of an ephemeral
```

File: website/docs/language/resources/ephemeral.mdx [hunk 28]
```diff
@@ -0,0 +1,137 @@
...
+resource in a given configuration.
+
+### Dependency graph
+
+Ephemeeral resources form nodes in Terraform's dependency graph which interact
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:84]:
```suggestion
Ephemeral resources form nodes in Terraform's dependency graph which interact
```

File: website/docs/language/resources/ephemeral.mdx [hunk 29]
```diff
@@ -0,0 +1,137 @@
...
+similarly to resources and data sources. For example, when a resource or data
+source depends on an attribute of an ephemeral resource, Terraform will
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:86]:
```suggestion
source depends on an attribute of an ephemeral resource, Terraform
```

File: website/docs/language/resources/ephemeral.mdx [hunk 30]
```diff
@@ -0,0 +1,137 @@
...
+automatically provision the ephemeral resource first.
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:87]:
```suggestion
automatically provisions the ephemeral resource first.
```

File: website/docs/language/resources/ephemeral.mdx [hunk 31]
```diff
@@ -0,0 +1,137 @@
...
+
+## Configuration model
+
+The following list outlines field hierarchy, language-specific data types, and
+requirements in the `moved` block.
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:92]:
```suggestion
requirements in the `ephemeral` block. 
```

File: website/docs/language/resources/ephemeral.mdx [hunk 32]
```diff
@@ -0,0 +1,137 @@
...
+An `ephemeral` block has the following form:
+
+```hcl
+ephemeral "<resource_type>" "<resource_name>" {
+  <attributes>
+  <meta-arguments>
+}
+```
+
+## Specification
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:109]: Maybe just remove this specification section for now? I understand it's tricky with a resource and I don't really think we need to figure this out for the beta

**Review by rkoron007 — APPROVED** [review]
Looking great so far! Left style comments, feel free to push back on anything that doesn't feel good!

File: website/docs/language/values/locals.mdx [hunk 33]
```diff
@@ -54,6 +54,37 @@ locals {
 }
 ```
 
+## Ephemeral values
+
```
Comment (rkoron007) [inline website/docs/language/values/locals.mdx:58]: Can you add a beta warning here too?

File: website/docs/language/resources/ephemeral.mdx [hunk 34]
```diff
@@ -0,0 +1,137 @@
...
+
+```hcl
+provider random {}
+
+ephemeral "random_password" "password" {
+  length           = 16
+  special          = true
+  override_special = "!#$%&*()-_=+[]{}<>:?"
+}
+```
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:137]: It'd also be super nice to have a more complicated example, like setting up a provider configuration with an ephemeral resource, [like in this demo](https://hashicorp.slack.com/archives/C071HC4JJCC/p1729859992489559). If it wouldn't be too much work I think it'd really be the bread and butter of this article if you can pull it off

File: website/docs/language/values/locals.mdx [hunk 35]
```diff
@@ -54,6 +54,37 @@ locals {
...
+Local values will implicitly become ephemeral if an ephemeral value is
+referenced in their assignment. For example:
+
+```hcl
+variable "service_name" {
+	type    = string
```
Comment (rkoron007) [inline website/docs/language/values/locals.mdx:64]: formatting on this one

File: website/docs/language/resources/ephemeral.mdx [hunk 36]
```diff
@@ -0,0 +1,137 @@
...
+
+This topic provides reference information for the `moved` block. 
+
+## Introduction
+
+_Ephemeral Resources_ are Terraform resources that are temporary, not stored in
+state, and have a unique lifecycle. Each ephemeral block describes one or more
+ephemeral resources, such as a temporary password or connection to another
+system.
+
```
Comment (rkoron007) [inline website/docs/language/resources/ephemeral.mdx:18]: Can we add a list somewhere for where you can use ephemeral resources? Something like this:

```
You can only reference ephemeral resources in specific ephemeral contexts or Terraform throws an error. The following are valid contexts for referencing ephemeral resources:
* In another ephemeral resource
* In [local values](/terraform/language/values/locals#ephemeral-values)
* In [ephemeral variables](terraform/language/values/variables#exclude-values-from-state)
* In [ephemeral outputs](/terraform/language/values/outputs#ephemeral-avoid-storing-values-in-state-or-plan-files)
```

Comment (robin-norwood) [reply website/docs/language/resources/ephemeral.mdx:18]: Check where I added it, not sure where else?

**Pull request merged**